bansa/src/main/kotlin/com/brianegan/bansa/reselect.kt

```diff
@@ -0,0 +1,234 @@
+package com.brianegan.bansa
+
+/**
+ * A rewrite for kotlin of https://github.com/reactjs/reselect library for redux (https://github.com/reactjs/redux)
+ * see also "Computing Derived Data" in redux documentation http://redux.js.org/docs/recipes/ComputingDerivedData.html
+ * Created by Dario Elyasy  on 3/18/2016.
+ */
+private val defaultEqualityCheck = { a: Any, b: Any -> a === b }
+
+interface Memoizer<T> {
+    fun memoize(vararg inputArgs: Any): T
+}
+
+inline fun <T : Any> Array<out T>.every(transform: (Int, T) -> Boolean): Boolean {
+    forEachIndexed { i, t -> if (!transform(i, t)) return false }
+    return true
+}
+
+
+// {a:Any,b:Any -> a===b}
+fun <T> defaultMemoize(func: (Array<out Any>) -> T, equalityCheck: (a: Any, b: Any) -> Boolean = defaultEqualityCheck) = object : Memoizer<T> {
+    var lastArgs: Array<out Any>? = null
+    var lastResult: T? = null
+    override fun memoize(vararg inputArgs: Any): T {
+        if (lastArgs != null &&
+                lastArgs!!.size == inputArgs.size && inputArgs.every { index, value -> equalityCheck(value, lastArgs!![index]) }) {
+            return lastResult!!
+        }
+        lastArgs = inputArgs
+        lastResult = func(inputArgs)
+        return lastResult!!
+    }
+}
+
+interface SelectorInput<S, I> {
+    operator fun invoke(state: S): I
+}
+
+/**
+ * a selector function is a function that map a field in state object to the input for the selector compute function
+ */
+class InputField<S, I>(val fn: S.() -> I) : SelectorInput<S, I> {
+    override operator fun invoke(state: S): I = state.fn()
+}
+
+
+/**
+ * note: [Selector] inherit from [SelectorInput] because of support for composite selectors
+ */
+interface Selector<S, O> : SelectorInput<S, O> {
+    val recomputations: Long
+    fun isChanged(): Boolean
+    fun resetChanged()
+    fun getIfChangedIn(state: S): O? {
+        var res = invoke(state)
+        if (isChanged()) {
+            resetChanged()
+            return res
+        }
+        return null
+    }
+
+    fun onChangeIn(state: S, blockfn: (O) -> Unit) {
+        getIfChangedIn(state)?.let(blockfn)
+    }
+}
+
+
+/**
+ * abstract base class for all selectors
+ */
+abstract class AbstractSelector<S, O> : Selector<S, O> {
+    protected var recomputationsLastChanged = 0L
+    protected var _recomputations = 0L
```

Comment:
Ack, yah, I should have looked at this closer. Just a heads up: you could totally ditch `_recomputations` here and simply use:

```kotlin
override var recomputations = 0L
```